pkg/networkservice/common/mechanisms/kernel/client.go

```diff
@@ -44,23 +46,40 @@ func NewClient(options ...Option) networkservice.NetworkServiceClient {
 }
 
 func (k *kernelMechanismClient) Request(ctx context.Context, request *networkservice.NetworkServiceRequest, opts ...grpc.CallOption) (*networkservice.Connection, error) {
-	// Append the mechanism preference to the request
-	preferredMechanism := &networkservice.Mechanism{
-		Cls:  cls.LOCAL,
-		Type: kernel.MECHANISM,
-		Parameters: map[string]string{
-			kernel.NetNSURL:         (&url.URL{Scheme: "file", Path: netNSFilename}).String(),
-			kernel.InterfaceNameKey: k.interfaceName,
-		},
+	if !updateMechanismPreferences(request) {
+		request.MechanismPreferences = append(request.GetMechanismPreferences(), &networkservice.Mechanism{
+			Cls:  cls.LOCAL,
+			Type: kernel.MECHANISM,
+			Parameters: map[string]string{
+				kernel.NetNSURL:         (&url.URL{Scheme: "file", Path: netNSFilename}).String(),
+				kernel.InterfaceNameKey: k.interfaceName,
+			},
+		})
 	}
-	request.MechanismPreferences = append(request.GetMechanismPreferences(), preferredMechanism)
 	return next.Client(ctx).Request(ctx, request, opts...)
 }
 
 func (k *kernelMechanismClient) Close(ctx context.Context, conn *networkservice.Connection, opts ...grpc.CallOption) (*empty.Empty, error) {
 	return next.Client(ctx).Close(ctx, conn, opts...)
 }
 
+// updateMechanismPreferences returns true if MechanismPreferences has updated
+func updateMechanismPreferences(request *networkservice.NetworkServiceRequest) bool {
+	var updated = false
+
+	for _, m := range request.GetRequestMechanismPreferences() {
+		if m.Type == kernel.MECHANISM {
+			if m.Parameters == nil {
+				m.Parameters = make(map[string]string)
+			}
+			m.Parameters[kernel.NetNSURL] = (&url.URL{Scheme: "file", Path: netNSFilename}).String()
```

Comment:
Should we add kernel.InterfaceNameKey with k.interfaceName?

Reply:
yes, we should, fixed.